File: src/deque.py
"""Deque data structure."""
from doubly_linked_list import DoublyLinked
from que_ import Q


class Deque(Q):
    """Create eque data structure."""

    def __init__(self):
        """Create instance of Deque class."""
        self.deque = DoublyLinked()

    def append(self, val):
        """Add value to end of deque."""
        self.deque.append(val)

    def append_left(self, val):
        """Add val to front of deque."""
        self.deque.push(val)

    def pop(self):
        """Remove and return value of deque end."""
        return self.deque.shift()

    def pop_left(self):
        """Remove and return value of deque front."""
        return self.deque.pop()

    def peek_left(self):
        """Return next value in deque."""
        return super(Q, self).peek()

    def peek(self):
        """Return tail value in deque."""
        if len(self.queue) > 0:
            return self.deque.tail.data
        return None

    def size(self):
        """Return length of deque."""
        return self.deque._counter

File: js/deque.js
'use strict';

let DoublyLinked = require('./doubly-linked')

class Deque {
  constructor() {
    this.deque = new DoublyLinked();
  }

  append(val) {
    this.deque.append(val);
  }

  appendLeft(val) {
    this.deque.push(val);
  }

  pop() {
    if (this.deque._counter == 0) {
      throw new Error('Cannot pop from empty deque.');
    }
    return this.deque.shift();
  }

  popLeft() {
    if (this.deque._counter == 0) {
      throw new Error('Cannot pop from empty deque.');
    }
    return this.deque.pop();
  }

  peek() {
    if(this.deque._counter == 0){
      return null;
    }
    return this.deque.tail.data;
  }

  peekLeft(){
    if (this.deque._counter == 0) {
      return null;
    }
    return this.deque.head.data;
  }
  // def peek_left(self):
  //       """Return next value in deque."""
  //       if len(self.deque) > 0:
  //           return self.deque.head.data
  //       return None
  size() {
    return this.deque._counter;
  }
}

module.exports = Deque
